src/history/hs.c

```diff
@@ -824,179 +823,166 @@ __wt_find_hs_upd(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_UPDATE **upd
     ret = __wt_hs_cursor_position(session, hs_cursor, hs_btree_id, key, read_timestamp);
-    for (; ret == 0; ret = hs_cursor->prev(hs_cursor)) {
-        WT_ERR(hs_cursor->get_key(hs_cursor, &hs_btree_id, hs_key, &hs_start.timestamp,
-          &hs_start.txnid, &hs_stop.timestamp, &hs_stop.txnid));
-
-        /* Stop before crossing over to the next btree */
-        if (hs_btree_id != S2BT(session)->id)
-            break;
+    if (ret == WT_NOTFOUND) {
+        ret = 0;
+        goto done;
+    }
+    WT_ERR(ret);
+    WT_ERR(hs_cursor->get_key(hs_cursor, &hs_btree_id, hs_key, &hs_start_ts, &hs_counter));
 
-        /*
-         * Keys are sorted in an order, skip the ones before the desired key, and bail out if we
-         * have crossed over the desired key and not found the record we are looking for.
-         */
-        WT_ERR(__wt_compare(session, NULL, hs_key, key, &cmp));
-        if (cmp != 0)
-            break;
+    /* Stop before crossing over to the next btree */
+    if (hs_btree_id != S2BT(session)->id)
+        goto done;
 
-        /*
-         * It is safe to assume that we're reading the updates newest to the oldest. We can quit
-         * searching after finding the newest visible record.
-         */
-        if (!__wt_txn_visible(session, hs_start.txnid, hs_start.timestamp))
-            continue;
+    /*
+     * Keys are sorted in an order, skip the ones before the desired key, and bail out if we have
+     * crossed over the desired key and not found the record we are looking for.
+     */
+    WT_ERR(__wt_compare(session, NULL, hs_key, key, &cmp));
+    if (cmp != 0)
+        goto done;
 
-        WT_ERR(
-          hs_cursor->get_value(hs_cursor, &durable_timestamp, &prepare_state, &upd_type, hs_value));
+    WT_ERR(hs_cursor->get_value(
+      hs_cursor, &hs_stop_ts, &durable_timestamp, &prepare_state, &upd_type, hs_value));
 
-        /* We do not have prepared updates in the history store anymore */
-        WT_ASSERT(session, prepare_state != WT_PREPARE_INPROGRESS);
+    /* We do not have prepared updates in the history store anymore */
+    WT_ASSERT(session, prepare_state != WT_PREPARE_INPROGRESS);
 
-        /*
-         * Found a visible record, return success unless it is prepared and we are not ignoring
-         * prepared.
-         *
-         * It's necessary to explicitly signal a prepare conflict so that the callers don't fallback
-         * to using something from the update list.
-         *
-         * FIXME-PM-1521: review the code in future
-         */
-        if (prepare_state == WT_PREPARE_INPROGRESS &&
-          !F_ISSET(&session->txn, WT_TXN_IGNORE_PREPARE) && !allow_prepare) {
-            ret = WT_PREPARE_CONFLICT;
-            break;
-        }
-
-        /* We do not have tombstones in the history store anymore. */
-        WT_ASSERT(session, upd_type != WT_UPDATE_TOMBSTONE);
+    /*
+     * Found a visible record, return success unless it is prepared and we are not ignoring
```

Comment:
We should never have an active prepared update in the history store (i.e. we should never have an update with prepare_state == WT_PREPARE_INPROGRESS in history store.